packages/app/src/cli/services/dev/extension.ts

```diff
@@ -109,32 +111,64 @@ export interface ExtensionDevOptions {
-  const payloadStoreOptions = {
+  const payloadOptions = {
     ...options,
     websocketURL: getWebSocketUrl(options.url),
   }
-  const bundlePath = options.appWatcher.buildOutputPath
-  const payloadStoreRawPayload = await getExtensionsPayloadStoreRawPayload(payloadStoreOptions, bundlePath)
-  const payloadStore = new ExtensionsPayloadStore(payloadStoreRawPayload, payloadStoreOptions)
 
-  outputDebug(`Setting up the UI extensions HTTP server...`, options.stdout)
-  const httpServer = setupHTTPServer({devOptions: options, payloadStore})
+  // NOTE: Always use `payloadOptions`, never `options` directly. This way we can mutate `payloadOptions` without
+  // affecting the original `options` object and we only need to care about `payloadOptions` in this function.
 
-  outputDebug(`Setting up the UI extensions Websocket server...`, options.stdout)
-  const websocketConnection = setupWebsocketConnection({...options, httpServer, payloadStore})
-  outputDebug(`Setting up the UI extensions bundler and file watching...`, options.stdout)
+  const bundlePath = payloadOptions.appWatcher.buildOutputPath
+  const payloadStoreRawPayload = await getExtensionsPayloadStoreRawPayload(payloadOptions, bundlePath)
+  const payloadStore = new ExtensionsPayloadStore(payloadStoreRawPayload, payloadOptions)
+
+  outputDebug(`Setting up the UI extensions HTTP server...`, payloadOptions.stdout)
+  const httpServer = setupHTTPServer({devOptions: payloadOptions, payloadStore})
+
+  outputDebug(`Setting up the UI extensions Websocket server...`, payloadOptions.stdout)
+  const websocketConnection = setupWebsocketConnection({...payloadOptions, httpServer, payloadStore})
+  outputDebug(`Setting up the UI extensions bundler and file watching...`, payloadOptions.stdout)
 
   const eventHandler = async ({extensionEvents}: AppEvent) => {
     for (const event of extensionEvents) {
       const status = event.buildResult?.status === 'ok' ? 'success' : 'error'
-      // eslint-disable-next-line no-await-in-loop
-      await payloadStore.updateExtension(event.extension, options, bundlePath, {status})
+
+      switch (event.type) {
+        case EventType.Created:
+          payloadOptions.extensions.push(event.extension)
+
+          if (!payloadOptions.checkoutCartUrl) {
+            // If the checkoutCartUrl is not set, check again and build the URL in case the new extension is a checkout extension.
+            try {
+              const cartUrl = await buildCartURLIfNeeded(payloadOptions.extensions, payloadOptions.storeFqdn)
+              // eslint-disable-next-line require-atomic-updates
+              payloadOptions.checkoutCartUrl = cartUrl
+              // eslint-disable-next-line no-catch-all/no-catch-all
+            } catch (error) {
+              outputDebug(
+                `Failed to build a cart URL for your checkout extension. Use the --checkout-cart-url flag to set a fixed URL.`,
+                payloadOptions.stdout,
+              )
```

Comment:
It seems like this is a unique error case because in the general case, if something fails inside `buildCartURLIfNeeded`, we can just crash dev and no harm done. I'm curious why we wouldn't:

1. Crash `dev` in this case
2. Keep going and just not provide the convenience link

Also, why would it not work? I guess just because:

1. Connection to the store fails
2. There are no products to test on

These feel like rare cases. 1 would be very odd given that we have other polling processes, and 2 would be odd not to crash or at least warn that no products exist. Note that the error message is raised inside `fetchProductVariant` so I'd expect we should at least display it.